### PR TITLE
Cell merging

### PR DESCRIPTION
**Pull request by ibdafna**
This PR introduces initial cell merging support for the lumino DataGrid. Immense thanks to @mbektasbbg for his mentorship and support during this project - it would not be possible without him.

lumino_merging_en

To maintain consistency with the existing data model API, groupCount() and group() methods have been added to the data model ABC. Here is an example of an implementing class:
```typescript
  groupCount(region: DataModel.RowRegion): number {
    if (region === 'body') {
      return 3;
    } else if (region === 'column-header') {
      return 1;
    } else if (region === 'row-header') {
      return 2;
    } else if (region === 'corner-header') {
      return 1;
    }
    return 0;
  }

  group(region: DataModel.CellRegion, groupIndex: number): CellGroup | null {
    if (region === 'body') {
      return [{startRow: 1, startColumn: 1, endRow: 2, endColumn: 2},
              {startRow: 5, startColumn: 1, endRow: 5, endColumn: 2},
              {startRow: 3, startColumn: 5, endRow: 4, endColumn: 5}][groupIndex];
    }

    if (region === 'column-header') {
      return [{startRow: 0, startColumn: 4, endRow: 1, endColumn: 4}][groupIndex];
    }

    if (region === 'row-header') {
      return [{startRow: 0, startColumn: 0, endRow: 1, endColumn: 1},
              {startRow: 4, startColumn: 0, endRow: 5, endColumn: 0}][groupIndex];
    }

    if (region === 'corner-header') {
      return [{startRow: 0, startColumn: 0, endRow: 1, endColumn: 1}][groupIndex];
    }

    return null;
  }
```

There are still some issues which need resolving, and we will be working on these in the upcoming weeks:

- Minor pixelation or lack of anti-aliasing of text when scrolling.
- Dragging row/column-header selections not always selecting the expected area.
- Discuss and decide on logic for how rows/columns should be highlighted when a merged cell region in the body is selected, but the row/column headers are not merged.
- Border rendering issue when a merged and selected region spanning beyond the visible grid area, is scrolled into the view.

Although still a work in progress, the PR is open for review. We welcome any comments or questions you have!

### CODE REVIEW

**Comment by ibdafna** [comment]
Addressing #103

**Comment by ibdafna** [comment]
@blink1073 @jasongrout @afshin for awareness

**Comment by bollwyvl** [comment]
@ibdafna this is _magic_ :sparkler: :unicorn: :rainbow: . What do you think needs to happen to un-draft this?

If this existed, I'd really love build a CollapsibleTreeModel on top of it that would be capable of doing file tree browsers, JSON viewers, etc.

A minor thing: perhaps this could be harmonized with the selection model, which has the same shape? `c1`, `r1` etc is a bit nastier to read, but easier to write :blush:

**Comment by ibdafna** [comment]
> 
> 
> @ibdafna this is _magic_ 🎇 🦄 🌈 . What do you think needs to happen to un-draft this?
> 
> If this existed, I'd really love build a CollapsibleTreeModel on top of it that would be capable of doing file tree browsers, JSON viewers, etc.
> 
> A minor thing: perhaps this could be harmonized with the selection model, which has the same shape? `c1`, `r1` etc is a bit nastier to read, but easier to write 😊

@bollwyvl Thank you! the main reason I left it as a draft was due to some edge cases with scrolling that needed fixing. I was also expecting review comments which would potentially require design changes, so I thought it would make sense to implement those changes before un-drafting. I will rebase the branch and un-draft - any reviews are welcome! And I'm happy to harmonize with the selection model 😃

**Comment by ibdafna** [comment]
Rebased and marking as ready for review. There are still some minor issues, but those can be fixed as a part of the review process! If I can get someone to review, I'm happy to jump on a call to clarify anything, and generally get this over the line in the coming days/weeks.

**Comment by SylvainCorlay** [comment]
Pinging @blink1073 and @afshin you may find it interesting.

This is already in use in [ipydatagrid](https://github.com/bloomberg/ipydatagrid).

File: packages/datagrid/src/cellgroup.ts [hunk 1]
```diff
@@ -0,0 +1,389 @@
+import { DataModel } from "./datamodel";
+import { SectionList } from './sectionlist';
+
+export
```
Comment (blink1073) [inline packages/datagrid/src/cellgroup.ts:4]: Can you please add docstrings to all public items in this file?

File: packages/datagrid/src/datamodel.ts [hunk 2]
```diff
@@ -91,6 +101,10 @@ abstract class DataModel {
   metadata(region: DataModel.CellRegion, row: number, column: number): DataModel.Metadata {
     return DataModel.emptyMetadata;
   }
+  
+  group(region: DataModel.CellRegion, groupIndex: number): CellGroup | null {
```
Comment (blink1073) [inline packages/datagrid/src/datamodel.ts:105]: Missing docstring

File: packages/datagrid/src/datagrid.ts [hunk 3]
```diff
@@ -3280,7 +3292,7 @@ class DataGrid extends Widget {
       let x = hw + this._columnSections.offsetOf(c);
       this.paintContent(x, 0, vw - x, vh);
     } else if (delta < 0) {
-      this.paintContent(vw + delta, 0, -delta, vh);
+      this.paintContent(0, 0, vw, vh); // TODO: check this.
```
Comment (blink1073) [inline packages/datagrid/src/datagrid.ts:3295]: Still todo?

File: packages/datagrid/src/datagrid.ts [hunk 4]
```diff
@@ -4524,13 +4842,49 @@ class DataGrid extends Widget {
...
+          }
+        }
+
+        topCurrent += rgn.rowSizes[rIndex];
+        if (r === rgn.row) {
+          topCurrent -= (rgn.yMin - rgn.y); 
+        }
+      }
+
+      // TODO: decide whether to use rgn.yMax or y2 (defined above)
```
Comment (blink1073) [inline packages/datagrid/src/datagrid.ts:4873]: Still todo?

**Comment by ibdafna** [comment]
@blink1073 we've ironed out all the quirks - I think it's ready to be merged 😸 
@bollwyvl renamed the interface as per your suggestion!

CC @ivanov AKA "The Unblocker" (read in "The Undertaker" from WWF tone) thank you for making this a smooth ride with your help.

**Review by blink1073 — APPROVED** [review]
Excellent, thank you!

**Comment by blink1073** [comment]
https://github.com/jupyterlab/lumino/releases/tag/2021.5.20

**Pull request merged**